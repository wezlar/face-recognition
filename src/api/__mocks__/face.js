import { mockFindBestMatch, mockGetFullFaceDescription } from './mockFunctions'; 

module.exports = {
  loadModels: jest.fn(),
  getFullFaceDescription: jest.fn().mockResolvedValue(mockGetFullFaceDescription),
  createMatcher: jest.fn().mockResolvedValue({
    findBestMatch: jest.fn().mockReturnValue(mockFindBestMatch)
  }),
};