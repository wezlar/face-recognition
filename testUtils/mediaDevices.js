export const mediaDevices = {
  enumerateDevices: jest.fn().mockResolvedValue([
    {
      deviceId: 'default',
      groupId: '93f7124db4f938763012fa70d923a5832e6304d0bb0508c8a9283b6fb3e901fa',
      kind: 'audioinput',
      label: '',
    },
    {
      deviceId: 'a4d796b64c620eb6ac4c6fea8eb68d2a1edf765f9dcfa21a53f963e6ed5332d8',
      groupId: '23aa31907dfcec8059ab203aa6c0dbea57a5420f1c42fa94efdc5021041ee763',
      kind: 'videoinput',
      label: 'FaceTime HD, Camera',
    },
    {
      deviceId: 'default',
      groupId: '93f7124db4f938763012fa70d923a5832e6304d0bb0508c8a9283b6fb3e901fa',
      kind: 'audiooutput',
      label: '',
    },
  ]),
};

export const multipleCameras = {
  enumerateDevices: jest.fn().mockResolvedValue([
    {
      deviceId: 'a4d796b64c620eb6ac4c6fea8eb68d2a1edf765f9dcfa21a53f963e6ed5332d8',
      groupId: '23aa31907dfcec8059ab203aa6c0dbea57a5420f1c42fa94efdc5021041ee763',
      kind: 'videoinput',
      label: 'FaceTime HD, Camera',
    },
    {
      deviceId: 'default',
      groupId: '93f7124db4f938763012fa70d923a5832e6304d0bb0508c8a9283b6fb3e901fa',
      kind: 'videoinput',
      label: 'Extra test camera',
    },
  ]),
};
