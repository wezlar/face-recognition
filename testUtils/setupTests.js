import * as emotion from '@emotion/core';
import { createSerializer } from 'jest-emotion';

expect.addSnapshotSerializer(createSerializer(emotion));