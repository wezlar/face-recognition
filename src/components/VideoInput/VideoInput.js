import React, { Component } from 'react';
import Webcam from 'react-webcam';
import { loadModels, getFullFaceDescription, createMatcher } from '../../api/face';
import DrawBox from '../DrawBox/DrawBox';

import { JSON_PROFILE, INIT_STATE, VIDEO_SIZE } from '../../constants';

const inputSize = 160;

class VideoInput extends Component {
  _isMounted = false;

  constructor (props) {
    super(props);
    this.webcam = React.createRef();
    this.state = { ...INIT_STATE };
  };

  componentWillMount = async () => {
    await loadModels();
    this.setInputDevice();
  };

  async componentDidMount () {
    this._isMounted = true;
    const matchedProfile = await createMatcher(JSON_PROFILE);
    if (this._isMounted) {
      this.setState({ faceMatcher: matchedProfile });
    }
  }

  componentWillUnmount () {
    clearInterval(this.interval);
    this._isMounted = false;
  };

  setInputDevice = () => {
    if (!this._isMounted) return;
    navigator.mediaDevices.enumerateDevices().then(async devices => {
      let inputDevice = await devices.filter(
        device => device.kind === 'videoinput'
      );
      
      if (!inputDevice.length) {
        await this.setState({
          facingMode: 'none',
        });
      } else if (inputDevice.length < 2) {
        if (this._isMounted) {
          await this.setState({
            facingMode: 'user',
          });
        }
      } else {
        if (this._isMounted) {
          await this.setState({
            facingMode: { exact: 'environment' },
          });
        }
      }
      this.startCapture();
    });
  };

  startCapture = () => {
    if (!!this.webcam.current) {
      this.capture();
      this.interval = setInterval(() => {
        this.capture();
      }, 1500);
    }
  };

  capture = async () => {
    if (!this.webcam.current) return;
    await getFullFaceDescription(
      this.webcam.current.getScreenshot(),
      inputSize
    ).then(fullDesc => {
      if (!!fullDesc) {
        this.setState({
          detections: fullDesc.map(fd => fd.detection),
          descriptors: fullDesc.map(fd => fd.descriptor),
        });
      }
    });

    if (!!this.state.descriptors && !!this.state.faceMatcher) {
      let match = await this.state.descriptors.map(descriptor => {
        return this.state.faceMatcher.findBestMatch(descriptor);
      });
      this.setState({ match });
    }
  };

  render () {
    const { detections, match, facingMode } = this.state;
    let videoConstraints = null;
    let camera = '';
    
    if (!!facingMode) {
      videoConstraints = {
        width: VIDEO_SIZE.WIDTH,
        height: VIDEO_SIZE.HEIGHT,
        facingMode,
      };
      if (facingMode === 'none') {
        camera = 'none';
      } else if (facingMode === 'user') {
        camera = 'Front';
      } else {
        camera = 'Back';
      }
    }

    return (
      <div
        className='camera'
        style={{
          display: 'flex',
          flexDirection: 'column',
          alignItems: 'center',
        }} >
        <p>Camera: {camera}</p>
        <div
          style={{
            width: VIDEO_SIZE.WIDTH,
            height: VIDEO_SIZE.HEIGHT,
          }} >
          <div style={{ position: 'relative', width: VIDEO_SIZE.WIDTH }}>
            {!!videoConstraints && camera !== 'none' ? (
              <div data-testid='webcam' style={{ position: 'absolute' }}>
                <Webcam
                  audio={false}
                  width={VIDEO_SIZE.WIDTH}
                  height={VIDEO_SIZE.HEIGHT}
                  ref={this.webcam}
                  screenshotFormat='image/jpeg'
                  videoConstraints={videoConstraints} />
              </div>
            ) : <div data-testid='webcam' style={{ position: 'absolute' }}>No camera</div> }
            <DrawBox key={1} detections={detections} match={match} />
          </div>
        </div>
      </div>
    );
  };
};

export default VideoInput;