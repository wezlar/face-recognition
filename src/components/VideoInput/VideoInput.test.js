import React from 'react';
import {
  render,
  cleanup,
  waitForElement,
} from '@testing-library/react';

import VideoInput from './VideoInput';
import { loadModels, getFullFaceDescription, createMatcher } from '../../api/face';
import { mediaDevices, multipleCameras } from '../../../testUtils/mediaDevices';

const originalWindow = global.Window;
global.window = Object.create(window);

jest.mock('../../api/face');
beforeEach(() => {
  Object.defineProperty(window, 'navigator', {
    value: {
      mediaDevices,
    },
  });
});

afterEach(() => {
  jest.useRealTimers();
  jest.restoreAllMocks();
  global.window = originalWindow;
  cleanup();
});

describe('Test VideoInput', () => {
  test('Test VideoInput Basics', async () => {
    const { container, getByTestId } = render(<VideoInput />);
    expect(container.firstChild).toMatchSnapshot();
    
    const webcam = await waitForElement(() => getByTestId('webcam'));
    expect(webcam).toMatchSnapshot();
    const drawbox = await waitForElement(() => getByTestId('draw-box'));
    expect(drawbox).toMatchSnapshot();

    expect(window.navigator.mediaDevices.enumerateDevices).toHaveBeenCalledTimes(1);
    expect(loadModels).toHaveBeenCalledTimes(1);
    expect(getFullFaceDescription).toHaveBeenCalledTimes(1);
    expect(createMatcher).toHaveBeenCalledTimes(1);
  });

  test('The setInterval should fire', async () => {
    jest.useFakeTimers();
    
    const { container, getByTestId } = render(<VideoInput />);
    expect(container.firstChild).toMatchSnapshot();
    
    const webcam = await waitForElement(() => getByTestId('webcam'));
    jest.advanceTimersByTime(2000);
    expect(setInterval).toHaveBeenCalledTimes(1);
    expect(loadModels).toHaveBeenCalledTimes(1);
    expect(getFullFaceDescription).toHaveBeenCalledTimes(2);
    expect(createMatcher).toHaveBeenCalledTimes(1);
    expect(webcam).toMatchSnapshot();
  });

  test('Stops interval on unmount', async () => {
    jest.useFakeTimers();
    
    const { unmount } = render(<VideoInput />);
    unmount();
    jest.runOnlyPendingTimers();
    expect(setInterval).toHaveBeenCalledTimes(0);
    expect(loadModels).toHaveBeenCalledTimes(1);
    expect(getFullFaceDescription).toHaveBeenCalledTimes(0);
    expect(createMatcher).toHaveBeenCalledTimes(1);
  });

  test('Test difference multiple cameras', async () => {
    window.navigator.mediaDevices = multipleCameras;

    const { getByTestId } = render(<VideoInput />);
    const webcam = await waitForElement(() => getByTestId('webcam'));
    expect(window.navigator.mediaDevices.enumerateDevices).toHaveBeenCalledTimes(1);
    expect(loadModels).toHaveBeenCalledTimes(1);
    expect(getFullFaceDescription).toHaveBeenCalledTimes(1);
    expect(createMatcher).toHaveBeenCalledTimes(1);
  });

  test('Without any webcams', async () => {
    window.navigator.mediaDevices.enumerateDevices = jest.fn().mockResolvedValue([]);

    const { container } = render(<VideoInput />);
    expect(container.firstChild).toMatchSnapshot();
    expect(getFullFaceDescription).toHaveBeenCalledTimes(0);
  });

  test('Test getFullFaceDescription when returns empty fullDesc', async () => {
    window.navigator.mediaDevices = mediaDevices;
    getFullFaceDescription.mockResolvedValue();
    const { getByTestId } = render(<VideoInput />);

    await waitForElement(() => getByTestId('webcam'));
    expect(getFullFaceDescription).toHaveBeenCalledTimes(1);
  });
});
