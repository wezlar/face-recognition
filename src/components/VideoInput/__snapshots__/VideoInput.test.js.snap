// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test VideoInput Test VideoInput Basics 1`] = `
<div
  class="camera"
  style="display: flex; flex-direction: column; align-items: center;"
>
  <p>
    Camera: 
    
  </p>
  <div
    style="width: 520px; height: 420px;"
  >
    <div
      style="position: relative; width: 520px;"
    >
      <div
        data-testid="webcam"
        style="position: absolute;"
      >
        No camera
      </div>
    </div>
  </div>
</div>
`;

exports[`Test VideoInput Test VideoInput Basics 2`] = `
<div
  data-testid="webcam"
  style="position: absolute;"
>
  <video
    autoplay=""
    class=""
    height="420"
    playsinline=""
    width="520"
  />
</div>
`;

exports[`Test VideoInput Test VideoInput Basics 3`] = `
<div
  data-testid="draw-box"
  style="position: absolute; border: solid; border-color: blue; height: 156.0831506280933px; width: 293.60938969101437px; transform: translate(293.60938969101437px, 88.34867537841333px);"
>
  <p
    data-testid="draw-name-box"
    style="background-color: blue; border: solid; border-color: blue; margin-top: 0px; color: rgb(255, 255, 255); width: 293.60938969101437px; transform: translate(-3px, 156.0831506280933px);"
  >
    Rio
  </p>
</div>
`;

exports[`Test VideoInput The setInterval should fire 1`] = `
<div
  class="camera"
  style="display: flex; flex-direction: column; align-items: center;"
>
  <p>
    Camera: 
    
  </p>
  <div
    style="width: 520px; height: 420px;"
  >
    <div
      style="position: relative; width: 520px;"
    >
      <div
        data-testid="webcam"
        style="position: absolute;"
      >
        No camera
      </div>
    </div>
  </div>
</div>
`;

exports[`Test VideoInput The setInterval should fire 2`] = `
<div
  data-testid="webcam"
  style="position: absolute;"
>
  <video
    autoplay=""
    class=""
    height="420"
    playsinline=""
    width="520"
  />
</div>
`;

exports[`Test VideoInput Without any webcams 1`] = `
<div
  class="camera"
  style="display: flex; flex-direction: column; align-items: center;"
>
  <p>
    Camera: 
    
  </p>
  <div
    style="width: 520px; height: 420px;"
  >
    <div
      style="position: relative; width: 520px;"
    >
      <div
        data-testid="webcam"
        style="position: absolute;"
      >
        No camera
      </div>
    </div>
  </div>
</div>
`;
